pypesto/objective/finite_difference.py

```diff
@@ -1,13 +1,215 @@
 """Finite differences."""
 
-from typing import List, Tuple, Union
+from typing import Callable, List, Tuple, Union
 import numpy as np
+import logging
 
 from .base import ObjectiveBase, ResultDict
 from .constants import (
     MODE_FUN, MODE_RES, FVAL, GRAD, HESS, RES, SRES,
 )
 
+logger = logging.getLogger(__name__)
+
+
+class Delta:
+    """Finite difference step size with automatic updating.
+
+    Parameters
+    ----------
+    delta:
+        (Initial) step size, either a float, or a vector of size (n_par,).
+        If not None, this is used as initial step size.
+    test_deltas:
+        Step sizes to try out in step size selection. If None, a range
+        [1e-2, 1e-3, ..., 1e-8] is considered.
+    update_condition:
+        A "good" step size may be a local property. Thus, this class allows
+        updating the step size if certain criteria are met, in the
+        :func:`pypesto.objective.finite_difference.Delta.update`
+        function.
+        Delta.CONSTANT means that the step size is only initially selected.
+        Delta.SPACE means that the step size is updated if the current
+        evaluation point is sufficiently far away from the last training point.
+        Delta.STEPS means that the step size is updated `max_steps` evaluations
+        after the last update.
+    max_steps:
+        Number of steps after which to update in the `Delta.STEPS` update
+        condition.
+    """
+
+    # update conditions
+    CONSTANT = "constant"
+    SPACE = "space"
+    STEPS = "steps"
+    UPDATE_CONDITIONS = [CONSTANT, SPACE, STEPS]
+
+    def __init__(
+        self,
+        delta: Union[np.ndarray, float, None] = None,
+        test_deltas: np.ndarray = None,
+        update_condition: str = CONSTANT,
+        max_steps: int = 30,
+    ):
+        if not isinstance(delta, (np.ndarray, float)) and delta is not None:
+            raise ValueError(f"Unexpected type {type(delta)} for delta")
+        self.delta: Union[np.ndarray, float, None] = delta
+
+        if test_deltas is None:
+            test_deltas = np.array([10**(-i) for i in range(2, 9)])
```

Comment:
Maybe... one might want to include also -1 here, i.e., `range(1, 9)` rather than `range(2,9)`. If you don't want to increase cpu time, rather kick out the -8, i.e., go for `range(1,8)`